src/controllers/bookController.ts

```diff
@@ -0,0 +1,128 @@
+import { NextFunction, Request, Response } from "express";
+import Book, { IBook } from "../models/Book";
+import { StatusCodes } from "http-status-codes";
+import { BadRequestError, NotFoundError } from "../errors";
+import mongoose from "mongoose";
+
+// Interface for query parameters
+interface IGetBooksQuery {
+  search?: string;
+  categories?: string;
+  sort?: "a-z" | "z-a" | "latest" | "oldest";
+}
+
+// Get all books
+const getAllBooks = async (
+  req: Request<{}, {}, {}, IGetBooksQuery>,
+  res: Response,
+  next: NextFunction
+) => {
+  const { search, categories, sort } = req.query;
+  let query: any = {};
+
+  // Filter by categories
+  if (categories) {
```

Comment:
We should not allow someone to pass both a string or an array to us, we will have to remember that this could be multiple different data types everywhere else we use it going forwards. I'm not what data type Mongoose gives you (in Golang, every parameter is retrieved as a string, in which case this isArray check is not useful), but if we should make it clear to the users what we expect and send back an error if it's not the right type.